### ISSUE TITLE
twisted.test.test_tcp.LargeBufferTestCase.testWriter fails with iocp reactor

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3233]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#3233|
|Type|defect|
|Created|2008-05-12 19:50:55Z|
|Branch|https://github.com/twisted/twisted/tree/iocp-tcp-large-write-3233-2|

For example, on a recent buildot run:

```
[FAIL]: twisted.test.test_tcp.LargeBufferTestCase.testWriter

Traceback (most recent call last):
  File "c:\twistedbuildbot24\WXP32-full2.5-iocp\Twisted\twisted\test\test_tcp.py", line 1506, in check
    "(%d != %d)" % (clientF.len, self.datalen))
twisted.trial.unittest.FailTest: client didn't receive all the data it expected (0 != 62914560)
```

I think the fix is fairly trivial: `writeToHandle` doesn't limit the quantity of data sent.

<details><summary>Searchable metadata</summary>

```
trac-id__3233 3233
type__defect defect
reporter__therve therve
priority__highest highest
milestone__ 
branch__branches_iocp_tcp_large_write_3233_2 branches/iocp-tcp-large-write-3233-2
branch_author__pahan__therve pahan, therve
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1210621855000000 1210621855000000
changetime__1300332868000000 1300332868000000
version__None None
owner__PenguinOfDoom PenguinOfDoom
cc__ivank
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

(In [23607]) Branching to 'iocp-tcp-large-write-3233'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve removed owner|
|-|-|

Hooray for self-contained change! Please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @therve|
|-|-|

So simple looking. :)

Can you add better docs for `IWriteHandle.writeToHandle` and give the implementation a docstring as well?  In particular, it seems like the return value is somewhat important.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve removed owner|
|-|-|

I've added some doc. This is not totally clear to me as well, so I tried to explain what I understood.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @therve|
|-|-|

Cool.  Unfortunately, I still see some failures on some builders.  eg, on Vista:

```
===============================================================================
[FAIL]: twisted.test.test_tcp.LargeBufferTestCase.testWriter

Traceback (most recent call last):
  File "c:\twistedbuildbot\Vista-full2.5-scmikes-iocp\Twisted\twisted\test\test_tcp.py", line 1506, in check
    "(%d != %d)" % (clientF.len, self.datalen))
twisted.trial.unittest.FailTest: client didn't receive all the data it expected (62913428 != 62914560)
===============================================================================
```

The xp machines seem happier now though.  Do we have any idea what the difference might be?  If it's really fixed on xp, then I see no reason not to merge now; vista is not a supported platform.  I just wonder if it's really fixed on xp or if it's just getting lucky and some random irrelevant difference on vista is revealing a real problem that's obscured on xp.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|ivank commented|
|-|-|

I see this bug on Server 2008 SP1 (Python 2.6.2 32 bit, Twisted r27226). It seems to always fail, and does so with wildly varying numbers:

```
C:\temp\temp3>python C:\Python26\Scripts\trial.py -r iocp twisted.test.test_tcp.LargeBufferTestCase.testWriter
twisted.test.test_tcp
  LargeBufferTestCase
    testWriter ...                                                       [FAIL]

===============================================================================
[FAIL]: twisted.test.test_tcp.LargeBufferTestCase.testWriter

Traceback (most recent call last):
  File "C:\Python26\lib\site-packages\twisted\test\test_tcp.py", line 1544, in check
    "(%d != %d)" % (clientF.len, self.datalen))
twisted.trial.unittest.FailTest: client didn't receive all the data it expected (62907020 != 62914560)
-------------------------------------------------------------------------------
Ran 1 tests in 0.633s

FAILED (failures=1)

C:\temp\temp3>python C:\Python26\Scripts\trial.py -r iocp twisted.test.test_tcp.LargeBufferTestCase.testWriter
twisted.test.test_tcp
  LargeBufferTestCase
    testWriter ...                                                       [FAIL]

===============================================================================
[FAIL]: twisted.test.test_tcp.LargeBufferTestCase.testWriter

Traceback (most recent call last):
  File "C:\Python26\lib\site-packages\twisted\test\test_tcp.py", line 1544, in check
    "(%d != %d)" % (clientF.len, self.datalen))
twisted.trial.unittest.FailTest: client didn't receive all the data it expected (25256540 != 62914560)
-------------------------------------------------------------------------------
Ran 1 tests in 0.177s

Also:

twisted.trial.unittest.FailTest: client didn't receive all the data it expected (233600 != 62914560)
twisted.trial.unittest.FailTest: client didn't receive all the data it expected (16116372 != 62914560)
twisted.trial.unittest.FailTest: client didn't receive all the data it expected (2854300 != 62914560)
twisted.trial.unittest.FailTest: client didn't receive all the data it expected (9193620 != 62914560)
```

http://twistedmatrix.com/trac/ticket/3567

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|ivank commented|
|-|-|

Please ignore the last link to an irrelevant ticket in my previous comment.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|ivank commented|
|-|-|

sorry, I meant to Cc

**Comment by twisted-trac** [comment]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)<a name="note_9"></a>|@ghazel commented|
|-|-|

I'm not sure this problem is related to writes or write size at all. But slowing down or spacing out the writes might prevent the test from failing since the receiving side isn't reading quite as fast.

I did a little debugging of this test failure on Windows 7. Here's what I found:

The receiving side indicates a clean socket closure first, and then the sending side gets a RST - which indicates Windows thought there was pending IO on the socket.

iocpsupport claims WSARecv is returning 0 with a lpNumberOfBytesRecvd of 0. MSDN claims "For byte streams, zero bytes having been read (as indicated by a zero return value to indicate success, and lpNumberOfBytesRecvd value of zero) indicates graceful closure and that no more bytes will ever be read."

However, if I ignore this occurrence when WSARecv returns immediately (but continue to treat it as EOF when a callback presents this 0,0 result) I am able to continue reading bytes from the socket, and the test always passes.

I am much more inclined to think this is a mis-use of IOCP somewhere than a bug in WSARecv.

**Comment by twisted-trac** [comment]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)<a name="note_10"></a>|@ghazel commented|
|-|-|

Hmm, http://msdn.microsoft.com/en-us/library/aa365538(VS.85).aspx

**Comment by twisted-trac** [comment]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)<a name="note_11"></a>|@ghazel commented|
|-|-|

Ok, SetFileCompletionNotificationModes has core affinity problems and does not fix the problem here, so let's avoid that for now.

Looking at boost::asio, which seems to use IOCP successfully even at high speeds, the return value of WSARecv is only considered to complete an async operation if an error occurs.  Other completions, including EOF, are only processed when the overlapped is returned by GQCS.

So, iocpcreator should probably be modified to do the same. It certainly solves this bug.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="pahan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|pahan commented|
|-|-|

(In [31017]) Branching to 'iocp-tcp-large-write-3233-2'

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|PenguinOfDoom commented|
|-|-|

I removed immediate success handling from every operation

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set owner to PenguinOfDoom<br>@exarkun set status to `new`|
|-|-|

Woop.  Thanks for working on this.

  1. I'm not sure I fully understand the code.  You said you removed the immediate success handling, but there is lots of code that does `handleRead` if `rc != ERROR_IO_PENDING`.  That seems like what I would call "immediate success handling".  What's up?
  1. Could the `doRead` in udp.py go away and be replaced by a suitable implementation of `readFromHandle`, which the `doRead` in abstract.py will call?  Also why is `evt.buff` set to `self._readBuffers` in one place and `self._readBuffers[0]` in another?
  1. Some style issues here and there - stop using underscores in the middle of variables, add a few more docstrings (to changed methods) to explain what they do a little bit.
  1. There's the test that fails on POSIX that should be skipped, as we discussed IRL

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_17"></a>|PenguinOfDoom commented|
|-|-|

1. There is immediate success (rc == 0) and immediate failure (rc not in [0, ERROR_IO_PENDING]). The latter doesn't put an event into the completion queue and has to be handled immediately.
2. That code is, indeed, icky. I opened [[#4959](https://github.com/twisted/twisted/issues/4959)](https://github.com/twisted/twisted/issues/4959) to refactor it.
3. #4959 is probably a better place for mass renames.
4. The test is now skipped if the imports fail.

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|PenguinOfDoom set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_19"></a>|PenguinOfDoom commented|
|-|-|

http://xzrq.net/3233_coverage_1/

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun set owner to PenguinOfDoom|
|-|-|

Woop.  Hooray.  The implementation and test coverage looks good.  There are a couple minor issues with the test code itself:

  1. If you can remove the duplication of pauseProducing/resumeProducing in `StopStartReadingProtocol` methods without damaging readability too much, that'd be good.
  1. Some docstrings would be nice.  Fewer methods will let you write fewer docstrings, fwiw. ;)
  1. Rename the terrible methods if you like
  1. Use `assertTrue` instead of `assert_`, or use `assertEquals` if there's an `==` in the expression passed in.
  1. don't use `under_scores`, use `camelCase`
  1. Don't use semicolons!

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|PenguinOfDoom set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun set owner to PenguinOfDoom|
|-|-|

This looks sort of great now.  The gtk2reactor/glib2reactor failures need to be addressed.  It should be skipped if you don't want to figure out how it might be fixed.

**Comment by twisted-trac** [comment]
|<img alt="pahan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_23"></a>|pahan set status to `closed`|
|-|-|

(In [31217]) Merge iocp-tcp-large-write-3233-2

Author: PenguinOfDoom
Reviewer: exarkun
Fixes: [#3233](https://github.com/twisted/twisted/issues/3233)

Remove immediate successful completion handling for all overlapped operations.
